File: app/internal/publisher/template/java/LicenseEngine.java
package com.websmith.sdk;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import java.io.File;
import java.io.FileReader;
import java.nio.file.Path;
import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.Map;

public class LicenseEngine {
    private final Client client;
    private final CacheManager cache;
    private final String hardwareId;
    private JsonObject licenseData;
    private String licenseKey;

    public LicenseEngine() {
        this("config/api-config.json");
    }

    public LicenseEngine(String configPath) {
        JsonObject config = loadConfig(configPath);
        this.client = new Client(configPath);
        this.cache = new CacheManager(config);
        this.hardwareId = HardwareFingerprint.generate();
        this.licenseData = null;
        this.licenseKey = null;
    }

    private static JsonObject loadConfig(String path) {
        try {
            File f = new File(path);
            if (f.exists()) {
                try (FileReader reader = new FileReader(f)) {
                    return JsonParser.parseReader(reader).getAsJsonObject();
                }
            }
        } catch (Exception ignored) {}
        return new JsonObject();
    }

    public JsonObject validate(String licenseKey) throws Client.ApiException {
        this.licenseKey = licenseKey;
        String cachedStatus = (String) cache.get("license_status");
        if (cachedStatus != null) {
            try {
                JsonObject cached = JsonParser.parseString(cachedStatus).getAsJsonObject();
                if (cached.has("valid") && cached.get("valid").getAsBoolean()) {
                    this.licenseData = cached;
                    return cached;
                }
            } catch (Exception ignored) {}
        }
        JsonObject result = client.validateLicense(licenseKey, hardwareId);
        if (result.has("license")) {
            this.licenseData = result.getAsJsonObject("license");
        } else {
            this.licenseData = result;
        }
        if (result.has("valid") && result.get("valid").getAsBoolean()) {
            cache.set("license_status", this.licenseData.toString());
        }
        return result;
    }

    public JsonObject activate(String licenseKey, String deviceName) throws Client.ApiException {
        this.licenseKey = licenseKey;
        JsonObject result = client.activateLicense(licenseKey, hardwareId, deviceName);
        if (result.has("license")) {
            this.licenseData = result.getAsJsonObject("license");
        } else if (result.has("valid") && result.get("valid").getAsBoolean()) {
            this.licenseData = result;
        }
        if (result.has("valid") && result.get("valid").getAsBoolean()) {
            cache.set("license_status", this.licenseData.toString());
        } else {
            cache.remove("license_status");
        }
        return result;
    }

    public JsonObject deactivate(String licenseKey) throws Client.ApiException {
        String key = licenseKey != null ? licenseKey : this.licenseKey;
        if (key == null || key.isEmpty()) {
            throw new IllegalArgumentException("License key is required");
        }
        JsonObject result = client.deactivateLicense(key, hardwareId);
        this.licenseData = null;
        if (licenseKey == null || licenseKey.equals(this.licenseKey)) {
            this.licenseKey = null;
        }
        cache.remove("license_status");
        return result;
    }

    public JsonObject renew(String licenseKey, Integer extraDays) throws Client.ApiException {
        String key = licenseKey != null ? licenseKey : this.licenseKey;
        if (key == null || key.isEmpty()) {
            throw new IllegalArgumentException("License key is required");
        }
        JsonObject result = client.renewLicense(key, extraDays);
        if (result.has("license")) {
            this.licenseData = result.getAsJsonObject("license");
        }
        cache.remove("license_status");
        return result;
    }

    public JsonObject startTrial(String email, String customerName) throws Client.ApiException {
        return startTrial(email, customerName, null);
    }

    public JsonObject startTrial(String email, String customerName, JsonObject customerData) throws Client.ApiException {
        JsonObject result = client.startTrial(email, customerName, hardwareId, customerData);
        if (result.has("license")) {
            this.licenseData = result.getAsJsonObject("license");
        }
        cache.remove("license_status");
        return result;
    }

    public JsonObject checkTrial() throws Client.ApiException {
        JsonObject result = client.checkTrial(hardwareId);
        if (result.has("license")) {
            this.licenseData = result.getAsJsonObject("license");
        }
        return result;
    }

    public JsonObject convertTrial(String plan, String name, String email) throws Client.ApiException {
        JsonObject result = client.convertTrial(hardwareId, plan, name, email);
        if (result.has("license_key")) {
            this.licenseKey = result.get("license_key").getAsString();
        }
        if (result.has("license")) {
            this.licenseData = result.getAsJsonObject("license");
        }
        cache.remove("license_status");
        return result;
    }

    public JsonObject viewHardwareStatus() throws Client.ApiException {
        String currentHw = HardwareDetector.getHardwareId();
        JsonObject validateResult = validate();
        JsonObject data = validateResult.getAsJsonObject("data");
        String registeredHw = data != null && data.has("hardware_id") ? data.get("hardware_id").getAsString() : "";
        JsonObject result = new JsonObject();
        result.addProperty("matched", currentHw.equals(registeredHw));
        result.addProperty("current_hardware_id", currentHw);
        result.addProperty("registered_hardware_id", registeredHw);
        result.addProperty("message", "Hardware replacement requires administrator approval. Please contact support.");
        return result;
    }

    public JsonObject bindDevice(String licenseKey, String deviceName) throws Client.ApiException {
        String key = licenseKey != null ? licenseKey : this.licenseKey;
        if (key == null || key.isEmpty()) {
            throw new IllegalArgumentException("License key is required");
        }
        JsonObject result = client.bindDevice(key, hardwareId, deviceName);
        if (result.has("license")) {
            this.licenseData = result.getAsJsonObject("license");
        }
        cache.remove("license_status");
        return result;
    }

    public boolean hasLicenseKey() {
        return licenseKey != null && !licenseKey.isEmpty();
    }

    public boolean isValid() {
        if (licenseData == null) {
            String cached = (String) cache.get("license_status");
            if (cached != null) {
                try {
                    licenseData = JsonParser.parseString(cached).getAsJsonObject();
                } catch (Exception ignored) {}
            }
        }
        if (licenseData == null) return false;
        if (!licenseData.has("status")) return false;
        String status = licenseData.get("status").getAsString();
        if (!"active".equals(status)) return false;
        if (licenseData.has("expires_at") && !licenseData.get("expires_at").isJsonNull()) {
            try {
                String expiresAt = licenseData.get("expires_at").getAsString();
                Instant expiry = Instant.from(DateTimeFormatter.ISO_INSTANT.parse(expiresAt));
                if (expiry.isBefore(Instant.now())) return false;
            } catch (Exception ignored) {}
        }
        return true;
    }

    public JsonObject getLicenseInfo() {
        return licenseData;
    }
}

File: app/internal/publisher/template/dotnet/LicenseEngine.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Threading.Tasks;

namespace WebsmithSDK
{
    public class LicenseEngine
    {
        private readonly Client _client;
        private readonly CacheManager _cache;
        private readonly string _licenseKeyPath;
        private JsonElement _licenseData;
        private readonly Dictionary<string, object> _fingerprint;
        private string? _licenseKey;

        public LicenseEngine(Client client, string productId, int cacheTtlSeconds = 0)
        {
            _client = client;
            _cache = new CacheManager(productId, cacheTtlSeconds);
            _fingerprint = HardwareFingerprint.Generate();
            _licenseKeyPath = Path.Combine(
                Environment.GetEnvironmentVariable("HOME") ?? Environment.GetEnvironmentVariable("USERPROFILE") ?? ".",
                ".websmith", productId, "license_key.txt");
        }

        public void Initialize()
        {
        }

        public async Task<JsonDocument> Validate(string? licenseKey = null)
        {
            var key = licenseKey ?? _licenseKey;
            if (string.IsNullOrEmpty(key))
                throw new InvalidOperationException("No license key provided. Call Initialize() first or pass a key.");

            var deviceId = _fingerprint["fingerprint"]?.ToString() ?? "";
            var result = await _client.ValidateLicense(key, deviceId);
            if (result.RootElement.TryGetProperty("license", out var license))
                _licenseData = license;
            return result;
        }

        public async Task<JsonDocument> Activate(string licenseKey, string deviceName)
        {
            var deviceId = _fingerprint["fingerprint"]?.ToString() ?? "";
            var result = await _client.ActivateLicense(licenseKey, deviceId, deviceName);
            if (result.RootElement.TryGetProperty("license", out var license))
            {
                _licenseData = license;
                _licenseKey = licenseKey;
                await SaveLicenseKey(licenseKey);
            }
            return result;
        }

        public async Task<JsonDocument> Deactivate()
        {
            if (string.IsNullOrEmpty(_licenseKey))
                throw new InvalidOperationException("No license key loaded.");

            var deviceId = _fingerprint["fingerprint"]?.ToString() ?? "";
            var result = await _client.DeactivateLicense(_licenseKey, deviceId);
            if (File.Exists(_licenseKeyPath))
                File.Delete(_licenseKeyPath);
            _licenseKey = null;
            _licenseData = default;
            return result;
        }

        public async Task<JsonDocument> Renew()
        {
            if (string.IsNullOrEmpty(_licenseKey))
                throw new InvalidOperationException("No license key loaded.");

            var result = await _client.RenewLicense(_licenseKey);
            if (result.RootElement.TryGetProperty("license", out var license))
                _licenseData = license;
            return result;
        }

        public async Task<JsonDocument> StartTrial(string email, string customerName, Dictionary<string, object>? customerData = null)
        {
            return await _client.StartTrial(email, customerName, customerData);
        }

        public async Task<JsonDocument> CheckTrial()
        {
            var deviceId = _fingerprint["fingerprint"]?.ToString() ?? "";
            return await _client.CheckTrial(deviceId);
        }

        public async Task<JsonDocument> ConvertTrial(string plan, string customerName, string customerEmail)
        {
            var deviceId = _fingerprint["fingerprint"]?.ToString() ?? "";
            return await _client.ConvertTrial(deviceId, plan, customerName, customerEmail);
        }

        public async Task<JsonDocument> BindDevice(string licenseKey, string hardwareId, string deviceName)
        {
            var result = await _client.BindDevice(licenseKey, hardwareId, deviceName);
            if (result.RootElement.TryGetProperty("license", out var license))
                _licenseData = license;
            return result;
        }

        public Dictionary<string, object> ViewHardwareStatus()
        {
            var currentHw = _fingerprint["fingerprint"]?.ToString() ?? "";
            var validateResult = Validate().Result;
            if (validateResult.RootElement.TryGetProperty("data", out var dataEl) && dataEl.ValueKind == JsonValueKind.Object)
            {
                var registeredHw = dataEl.TryGetProperty("hardware_id", out var hwId) ? hwId.GetString() ?? "" : "";
                return new Dictionary<string, object>
                {
                    ["matched"] = currentHw == registeredHw,
                    ["current_hardware_id"] = currentHw,
                    ["registered_hardware_id"] = registeredHw,
                    ["message"] = "Hardware replacement requires administrator approval. Please contact support."
                };
            }
            return new Dictionary<string, object>
            {
                ["matched"] = false,
                ["current_hardware_id"] = currentHw,
                ["registered_hardware_id"] = "",
                ["message"] = "Hardware replacement requires administrator approval. Please contact support."
            };
        }

        public bool HasLicenseKey()
        {
            return !string.IsNullOrEmpty(_licenseKey);
        }

        public bool IsValid()
        {
            if (_licenseData.ValueKind == JsonValueKind.Undefined) return false;
            if (!_licenseData.TryGetProperty("status", out var status) || status.GetString() != "active") return false;
            if (_licenseData.TryGetProperty("expires_at", out var expiresAt) && expiresAt.ValueKind == JsonValueKind.String)
            {
                if (DateTime.TryParse(expiresAt.GetString(), out var exp) && exp < DateTime.UtcNow) return false;
            }
            return true;
        }

        public JsonElement GetLicenseInfo()
        {
            return _licenseData;
        }

        private async Task SaveLicenseKey(string key)
        {
            var dir = Path.GetDirectoryName(_licenseKeyPath);
            if (!string.IsNullOrEmpty(dir))
                Directory.CreateDirectory(dir);
            await File.WriteAllTextAsync(_licenseKeyPath, key);
        }
    }
}
